### PR TITLE
feat(router): add CanDeactivate next state argument

### LINKED ISSUE #9853: [Router 3] CanDeactivate target route
**I'm submitting a ...**  (check one with "x")

```
[ ] bug report
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When routing is about to happen, CanDeactivate guards can only know about component instance about to be deactivated/reused along with current route state of the application.

**Expected/desired behavior**
Add an extra parameter to `CanDeactivate` interface - something like `nextUrl` like so:

``` typescript
export interface CanDeactivate<T> {
    canDeactivate(component: T, route: ActivatedRouteSnapshot, state: RouterStateSnapshot, nextUrl: string): Observable<boolean> | boolean;
}
```

Using `nextUrl` instead of simple boolean value such as `willReuse` could help further and mimic the same functionality as deprecated router.

**What is the motivation / use case for changing the behavior?**
This will help differentiate if the component will be deactivated or reused when the url changes. If the component is going to be destroyed, we can postpone the navigation as needed. This could be cleaning up some data (which will be unnecessary if the component is going to be reused and lifecycle hooks can be meaningless if this task runs on a service instead of component itself) or simply running some animations (which can be used page transitions; `ngOnDestroy()` is run after component is detached and it does not wait for any observable like `CanDeactivate` anyways).

**Please tell us about your environment:**
- **Angular version:** 2.0.0-rc.4 - @angular/router 3.0.0-beta.2
- **Browser:** [all ] 
- **Language:** [ TypeScript ]

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Closes #9853

### CODE REVIEW

File: modules/@angular/router/src/interfaces.ts [hunk 1]
```diff
@@ -186,7 +186,8 @@ export interface CanActivateChild {
  *   canDeactivate(
  *     component: TeamComponent,
  *     route: ActivatedRouteSnapshot,
```
Comment (vicb) [inline modules/@angular/router/src/interfaces.ts]: currentRoute

File: modules/@angular/router/test/integration.spec.ts [hunk 2]
```diff
@@ -1624,7 +1625,9 @@ describe('Integration', () => {
             TestBed.configureTestingModule({
               providers: [{
                 provide: 'CanDeactivate',
-                useValue: (c: TeamCmp, a: ActivatedRouteSnapshot, b: RouterStateSnapshot) => {
+                useValue: (c: TeamCmp, a: ActivatedRouteSnapshot, b: RouterStateSnapshot,
```
Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: consistency `b` vs `nextState`

**Comment by vicb** [comment]
I'm not very clear on the use case but this is what should be tested in the integration test.

**Comment by DzmitryShylovich** [comment]
@vicb 
> should be tested in the integration test.

done

File: modules/@angular/router/test/integration.spec.ts [hunk 3]
```diff
@@ -1590,6 +1590,65 @@ describe('Integration', () => {
              expect(location.path()).toEqual('/simple');
            })));
 
+        describe('next state', () => {
+          class ClassWithNextState implements CanDeactivate<TeamCmp> {
+            canDeactivate(
+                component: TeamCmp, currentRoute: ActivatedRouteSnapshot,
+                currentState: RouterStateSnapshot, nextState: RouterStateSnapshot): boolean {
+              expect(currentState.url).toEqual('/team/22');
```
Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: maybe use a `log` and assert the content in test test ?

**Comment by vicb** [comment]
Could you please:
- address the comment in the tests,
- add a description of what this feature is in the commit comment.

Thanks !

**Comment by DzmitryShylovich** [comment]
@vicb done

**Comment by vicb** [comment]
Could you please reword it as 

```
Adds a `nextState` argument to access the future url from `CanDeactivate`.

BEFORE:
    
    canDeactivate(component: T, route: ActivatedRouteSnapshot, state: RouterStateSnapshot)
        :Observable<boolean>|Promise<boolean>|boolean;

AFTER:

    canDeactivate(component: T, currentRoute: ActivatedRouteSnapshot, currentState: RouterStateSnapshot,
                             nextState?: RouterStateSnapshot): Observable<boolean>|Promise<boolean>|boolean;

```

**Comment by DzmitryShylovich** [comment]
@vicb could u reword it on merge? :)

**Comment by vicb** [comment]
I would but I'm not merging this week

**Comment by nheidloff** [comment]
This seems promising. I hope it will be merged soon. Thanks @DzmitryShylovich

**Comment by DzmitryShylovich** [comment]
@vicb 
> I would but I'm not merging this week

oke, I did it 😄

**Pull request merged**